src/ir_passes/bundle_elim.rs

```diff
@@ -0,0 +1,233 @@
+use std::collections::HashMap;
+
+use itertools::Itertools;
+
+use crate::ir::{
+    Access, Command, CompIdx, Component, Connect, Context, Ctx, DenseIndexInfo,
+    Expr, Foreign, InvIdx, Invoke, Liveness, MutCtx, Port, PortIdx, Range,
+    Time,
+};
+
+#[derive(Default)]
+pub struct BundleElim {
+    context: DenseIndexInfo<Component, HashMap<PortIdx, Vec<PortIdx>>>,
+}
+
+impl BundleElim {
+    fn port(&self, pidx: PortIdx, comp: &mut Component) -> Vec<PortIdx> {
+        let one = comp.add(Expr::Concrete(1));
+        let Port {
+            owner,
+            width,
+            live,
+            info,
+        } = comp.get(pidx).clone();
+
+        let Liveness { idx, len, range } = live;
+
+        let start = comp.get(range.start).clone();
+        let end = comp.get(range.end).clone();
+
+        let len = len.as_concrete(comp).unwrap();
+
+        // if we need to preserve external interface information, we can't have bundle ports in the signature.
+        if comp.src_info.is_some()
+            && matches!(owner, crate::ir::PortOwner::Sig { .. })
+        {
+            assert!(
+                len == 1,
+                "Bundle ports in the signature are not supported."
+            );
+
+            // need to preserve the original portidx here to save the source information.
+            return vec![pidx];
+        }
+
+        let ports = (0..len)
+            .map(|i| {
+                let binding: HashMap<_, _> = [(idx, i)].into();
```

Comment:
This is where the `Bind` and `Subst` structures come in handy: you don't have to allocate expensive HashMaps.